### ISSUE TITLE
$PYTHONPATH is not set in Python module under some circumstances

### ISSUE THREAD

**Issue by boegel** [hpcugent/easybuild-easyblocks#609]
cc @Silox, @JensTimmerman, please track this issue here

**Comment by Silox** [comment]
This issue occured while installing Python 3.4.3 with the intel 2015a toolchain on the Delcatty cluster of the HPC UGent with EasyBuild 2.1. The full debug installation log can be found at http://kelder.zeus.ugent.be/~silox/eb-lockfile-2.1.log. The failing part is:

```
== 2015-05-07 23:25:06,806 main.EB_Python DEBUG modulename found in self.options, using it: lockfile
== 2015-05-07 23:25:06,806 main.run DEBUG run_cmd: running cmd python -c "import lockfile" (in /tmp)
== 2015-05-07 23:25:07,301 main.run DEBUG cmd "python -c "import lockfile"" exited with exitcode 1 and output:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.6/site-packages/lockfile/__init__.py", line 310, in <module>
    import linklockfile as _llf
ImportError: No module named 'linklockfile'
 
== 2015-05-07 23:25:07,302 main.EB_Python WARNING Extension: lockfile failed to install, cmd 'python -c "import lockfile"' (stdin: None) output: Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.6/site-packages/lockfile/__init__.py", line 310, in <module>
    import linklockfile as _llf
ImportError: No module named 'linklockfile'
 
== 2015-05-07 23:25:07,302 main.EB_Python WARNING Sanity check for lockfile failed: lockfile failed to install, cmd 'python -c "import lockfile"' (stdin: None) output: Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.6/site-packages/lockfile/__init__.py", line 310, in <module>
    import linklockfile as _llf
ImportError: No module named 'linklockfile
``` 

It seems like the wrong lockfile is being imported (the one from the default Python 2.6 system installation) instead of the one from Python 3.4.3 which was just installed.

After the failed installation, `easybuild/software/Python/3.4.3-intel-2015a/lib/python3.4/site-packages/lockfile` does exist, when I type `python -c 'import lockfile'` in the directory above that one, it seems to be loading fine. 

It seems that the $PYTHONPATH is not set correctly as there is no `setenv $PYTHONPATH` on line 148439 in my log file.

**Comment by boegel** [comment]
log also available in https://gist.github.com/boegel/c02ed8d3037e5abfe23a

**Comment by JensTimmerman** [comment]
Error: blob is too big

**Comment by JensTimmerman** [comment]
Well, clearly some path's are being set wrongly during the sanity check here: 
When building in `/tmp/vsc40485/eb/install/software/Python/3.4.3-intel-2015a/lib/python3.4/site-packages/` `PYTHONPATH` is set wrong:
```
== 2015-05-11 14:38:20,502 main.environment INFO Environment variable PYTHONPATH set to /apps/gent/SL6/sandybridge/software/EasyBuild/2.1.0/lib/python2.6/site-packages (previously undefined)
== 2015-05-11 14:38:23,366 main.EB_Python DEBUG modulename found in self.options, using it: lockfile
== 2015-05-11 14:38:23,366 main.run DEBUG run_cmd: running cmd python -c "import lockfile" (in /dev/shm)
== 2015-05-11 14:38:23,441 main.run DEBUG cmd "python -c "import lockfile"" exited with exitcode 1 and output:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.6/site-packages/lockfile/__init__.py", line 310, in <module>
    import linklockfile as _llf
ImportError: No module named 'linklockfile'
```

**Comment by JensTimmerman** [comment]
The module is loaded, but pythonpath is not set!

**Comment by boegel** [comment]
`$PYTHONPATH` should get set when extensions are being installed using `PythonPackage`, see https://github.com/hpcugent/easybuild-easyblocks/blob/master/easybuild/easyblocks/generic/pythonpackage.py#L244

Maybe the `make_module_extra` stuff that comes from the extensions doesn't get included anymore in the generated module.

That is supposed to be taken care of via https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/framework/easyblock.py#L1466, but `run()` doesn't return anything for `PythonPackage`, so it seems `PythonPackage.make_module_extra` is never called...

**Comment by boegel** [comment]
As far as I can tell, this is an issue that has been there for months/years... It seems setting `$PYTHONPATH` is only needed in very specific circumstances, and that `import`s usually go check in `lib/python3.4/site-packages`, regardless of whether `$PYTHONPATH` is extended or not...

**Comment by JensTimmerman** [comment]
I don't think the pythonpath should be set for every extension, since they all get installed in the default location. The pythonpath should be set in the python easyblock, so it at least ends up in the module file, even for a bare python install.

**Comment by JensTimmerman** [comment]
Ok, I think I might have zeroed in on the issue here, 

When loading the EasyBuild module, the easybuild install get's prepended to the path, and so does the system python, it seems.

So the path doesn't get cleaned enough before the sanity check is ran, 
the reason Silox and me can reproduce this is because we do a `module load EasyBuild`, whereas in your account, and the other account I first used to test eb was probably not prepended, but appended to the path?

```
== 2015-05-11 16:25:28,634 main.run DEBUG cmd "python -c 'import sys; print(sys.executable)'" exited with exitcode 0 and output:
/tmp/vsc40485/eb/install/software/Python/3.4.3-intel-2015a/bin/python

== 2015-05-11 16:25:28,634 main.run DEBUG Using default regular expression: (?<![(,-]|\w)(?:error|segmentation fault|failed)(?![(,-]|\.?\w)
== 2015-05-11 16:25:28,634 main.environment INFO Environment variable PYTHONNOUSERSITE set to 1 (previously undefined)
== 2015-05-11 16:25:28,634 main.run DEBUG run_cmd: running cmd python -c 'import sys; print(sys.path)' (in /dev/shm/Python/3.4.3/intel-2015a/setuptools/setuptools-11.3.1)
== 2015-05-11 16:25:28,683 main.run DEBUG cmd "python -c 'import sys; print(sys.path)'" exited with exitcode 0 and output:
['', '/apps/gent/SL6/sandybridge/software/EasyBuild/2.1.0/lib/python2.6/site-packages/vsc_base-2.2.0-py2.6.egg', '/apps/gent/SL6/sandybridge/software/EasyBuild/2.1.0/lib/python2.6/site-packages/easybuild_framework-2.1.0-py2.6.egg', '/apps/gent/SL6/sandybridge/software/EasyBuild/2.1.0/lib/python2.6/site-packages/easybuild_easyblocks-2.1.0-py2.6.egg', '/apps/gent/SL6/sandybridge/software/EasyBuild/2.1.0/lib/python2.6/site-packages/easybuild_easyconfigs-2.1.0-py2.6.egg', '/apps/gent/SL6/sandybridge/software/EasyBuild/2.1.0/lib/python2.6/site-packages', '/usr/lib/python2.6/site-packages', '/tmp/vsc40485/eb/install/software/Python/3.4.3-intel-2015a/lib/python34.zip', '/tmp/vsc40485/eb/install/software/Python/3.4.3-intel-2015a/lib/python3.4', '/tmp/vsc40485/eb/install/software/Python/3.4.3-intel-2015a/lib/python3.4/plat-linux', '/tmp/vsc40485/eb/install/software/Python/3.4.3-intel-2015a/lib/python3.4/lib-dynload', '/tmp/vsc40485/eb/install/software/Python/3.4.3-intel-2015a/lib/python3.4/site-packages']
```

**Comment by boegel** [comment]
@JensTimmerman: that looks like the cause indeed, but I'm not sure I understand yet what's going on exactly here...

Maybe it's just an issue of the order in which `$PYTHONPATH` is defined, and the modules are (re)loaded, causing things to be in the wrong order.

**Comment by boegel** [comment]
@gppezzi is running into weird problems with this easyconfig too...

He noticed that only in this Python 3 easyconfig, the `exts_filter` parameter is defined, as such:

```python
exts_filter = ('PYTHONPATH="%(installdir)s:$PYTHONPATH python -c "import %(ext_name)s"', '')
```

None of the other Python 3 easyconfigs have this, and I see no reason to include it either.

But having this set like this shouldn't be causing problems like this either...

**Comment by boegel** [comment]
This may be fixed by https://github.com/hpcugent/easybuild-easyconfigs/pull/1765, any chance for a confirmation @Silox?

**Comment by JensTimmerman** [comment]
yep, this could very well be fixed by #1765, I'll test this later, good catch.

**Comment by boegel** [comment]
Don't thank me, thank @gppezzi

**Comment by boegel** [comment]
assuming fixed by #1765
